src/compiler/crystal/codegen/debug.cr

```diff
@@ -122,30 +121,47 @@ module Crystal
       # puts "Unsupported type for debugging: #{type} (#{type.class})"
     end
 
-    def declare_variable(var_name, var_type, alloca, target_def)
-      location = target_def.location
+    def declare_parameter(arg_name, arg_type, arg_no, alloca, location)
+      return unless location
+
+      debug_type = get_debug_type(arg_type)
+      return unless debug_type
+
+      scope = get_current_debug_scope(location)
+      return unless scope
+
+      file, dir = file_and_dir(location.filename)
+      file = di_builder.create_file(file, dir)
+
+      var = di_builder.create_parameter_variable scope, arg_name, arg_no, file, location.line_number, debug_type
+      expr = di_builder.create_expression(nil, 0)
+
+      insert_debug_declaration alloca, var, expr
+    end
+
+    def declare_variable(var_name, var_type, alloca, location)
       return unless location
 
       debug_type = get_debug_type(var_type)
       return unless debug_type
 
       scope = get_current_debug_scope(location)
       return unless scope
+
       file, dir = file_and_dir(location.filename)
       file = di_builder.create_file(file, dir)
 
-      var = di_builder.create_local_variable LLVM::DwarfTag::AutoVariable,
-        scope, var_name, file, location.line_number, debug_type
+      var = di_builder.create_auto_variable scope, var_name, file, location.line_number, debug_type
       expr = di_builder.create_expression(nil, 0)
 
-      {% begin %}
-      {% if LibLLVM::IS_36 || LibLLVM::IS_35 %}
-      declare = di_builder.insert_declare_at_end(alloca, var, expr, alloca_block)
-      {% else %}
+      insert_debug_declaration alloca, var, expr
+    end
+
+    def insert_debug_declaration(alloca, var, expr)
       declare = di_builder.insert_declare_at_end(alloca, var, expr, builder.current_debug_location, alloca_block)
-      {% end %}
-      builder.set_metadata(declare, @dbg_kind, builder.current_debug_location)
-      {% end %}
+      # TODO: This is redundant for LLVM >= 3.8, but currently we don't have access to builder from DIBuilder
```

Comment:
Yep! Moved the logic to set the debug location to the C++ extension for LLVM <= 3.6 as well.

I also refactored these methods to avoid code repetition.